### DISCUSSION TITLE
MLNCustomStyleLayer Swift examples for both Metal and OpenGL render

### DISCUSSION THREAD

**Discussion by thothai94** [maplibre/maplibre-native#2173] [Q&A]
Hello everyone,

I'm planning to crate example for MLNCustomStyleLayer of MapLibre native by add some 3D model (.obj, .dae... files) into StyleLayer. My current step is add an MLNCustomStyleLayer into MapView use Metal and OpenGL render  but MapView does not show my CustomLayer.

Can anyone help me solve below issues? Please take a look at *[My Github repo](https://github.com/thothai94/metal-render-map.git)*
### Current environments:
- Xcode 15
- MapLibre 6.2.0 (use Swift Package Manager)
### My issues: 
- iOS application always use OpenGL render backend
- My CustomStyleLayer does not show in MapView

**Comment by louwers** [comment]
We switched to Metal in MapLibre Native iOS 6.0.0.

OpenGL will no longer be supported on iOS going forward. If you need to use custom OpenGL layers your best course of action is to stick to 5.x.x until you get the change to redo them with Metal.

**Reply by thothai94** [reply]
Thanks alot for your information. 
I'm going to adjust and remove OpenGL out of my plan to use the newest version from MapLibre. 
By the way, please tell me how to configure Metal as a default renderer for MapLibre? I added MapLibre by Swift Package Manager and it always use OpenGL in MLNCustomStyleLayer (statement #if MLN_RENDER_BACKEND_METAL always return false)

**Reply by louwers** [reply]
The `#define` is just used for compiling. Once the library is compiled the `#defines` don' transfer.

Try using the simulator, if it's somewhat usable you know you are using Metal. :)

**Reply by thothai94** [reply]
Hi @louwers ,

I think MapLibre inside my project use OpenGL render because class extended by MLNCustomStyleLayer has no member 'rendererEncoder'

Can you tell my how to import MapLibre use Metal render? I tried SPM from *[https://github.com/maplibre/maplibre-gl-native-distribution.git](https://github.com/maplibre/maplibre-gl-native-distribution.git)* t exact version "6.2.0" and custom build from main branch but it didn't work.

**Reply by kylba4enko** [reply]
@thothai94 Hi, did you solve the problem? Looks like I have the same as you described with "has no member 'rendererEncoder'". Nobody can answer the question how to properly enable rendering via Metal
